LogWindow sometimes obscured.  Some samples added to .organ file with incorrect path.  Wave tremulant addition tweak.

kwhite-uottawa opened the pull request:
I've been very happily using GoOdf to quicky create .organ files to test problems with the recent GrandOrgue release.  Thanks @larspalo for this great program!

I experienced a few unexpected "gotchas" (for me at least!) while investigating https://github.com/GrandOrgue/grandorgue/issues/2004#issuecomment-2483899017

FWIW: I use a non-tiling window manager (cwm) that does not display window frames, etc.

* Raise() the LogWindow after calling Show().  Otherwise the log window might be obscured and the messages missed. Define a SHOWLOGWINDOW macro to wrap this.
* It is (probably) not necessary to raise the main window.  This might also obscure the log window. 
* Only call MakeRelativeTo() if we have an absolute path. removeBaseOdfPath() may not need to calculate a relative path if the filename does not start with a path separator.
* Log any files that cannot be found when reading a .organ file and mention that they will be excluded/deleted.
* When adding a Wave tremulant and only a single ignore Wave tremulant attack exists, set that attack to play when Wave tremulant is off. Log if this has been done.  This behaviour is enabled when the bool setSingle is true (the default).

larspalo commented:
Thanks for your interest in GoOdf and it's great that you think it's useful! Sorry that I've not been able to respond so quickly but we've been without current here (due to a snow storm and a loss of power) for a few days, so no internet connection either until now.

> It is (probably) not necessary to raise the main window. This might also obscure the log window.

It's done for the Msw port to function as expected. Further testing is needed to verify. But it's unlikely to change.

> Log any files that cannot be found when reading a .organ file and mention that they will be excluded/deleted.

This is very good! I'll definitely include this in one way or another.

> When adding a Wave tremulant and only a single ignore Wave tremulant attack exists, set that attack to play when Wave tremulant is off. Log if this has been done. This behaviour is enabled when the bool setSingle is true (the default).

This is something I'll need to investigate to secure the correct behaviour in every case expected.

kwhite-uottawa commented:
I did very casual testing on a Win10 VM. The main window still appeared without the Raise() call.  I am **not** a windows user though.

My logging of missing "DUMMY" or "REF:*" files is probably not helpful.  I've added a check for that in ff81740.

I also added support for a command line .organ file argument in 3f1d122.  I found myself typing "GoOdf organ.organ" the same as I would "GrandOrgue organ.organ" too many times.  It was a "nice to have."

larspalo commented:
> The main window still appeared without the Raise() call. I am not a windows user though.

It appears the first time the program is opened, but if I remember correctly, on subsequent loadings of .organ files the main window "disappeared" and was needed to be called back by clicking on it in the tray. I'm not really a windows user myself either, but others are.

> My logging of missing "DUMMY" or "REF:*" files is probably not helpful.

No, likely not. I think that checking for correctly existing REF: already exist - but I'm unsure about if any logging is done. DUMMY is always valid and cannot really be missing.

> I also added support for a command line .organ file argument

I don't mind that. A way to register GoOdf as a program to open .organ files on all platforms directly from a file manager would also be nice.

larspalo commented:
Another small thing is that I'll want to keep the On: [push] criteria in .github/workflows/release.yml so that every push also will result in an intermediate build that can be tested by anyone.

On a side note, it would be better if you did any work/changes in a separate branch for each feature and created a separate PRs to GoOdf master from each one.

I'm sorry that I have so little time right now to test this because I really thing some of the changes you're proposing would be very valuable indeed. Hopefully, I'll get to it soon! Thanks.

larspalo commented on src/GOODFFunctions.h:
```diff
@@ -62,7 +68,7 @@ namespace GOODF_functions {
 		wxString stringToReturn = path;
 		wxFileName fName = wxFileName(path);
-		if (fName.FileExists()) {
+		if (fName.FileExists() && stringToReturn.StartsWith(wxFILE_SEP_PATH)) {
```
I'm not certain here, but I don't think that the string starting with a file separator should be part of the condition as the removeBaseOdfPath is used for a few other things too, like when the write location of the .organ file would be another from where it was read. Cleaning out the file separator, if existing, is a later step but the earlier could potentially still be needed. I would need further testing to be sure though.

kwhite-uottawa replied:
Without this test, when loading a rank with attacks and release GoOdf writes a .organ file that contains Pipe999Release999=`<path>` where path begins with "`../somewhere`" and  "`../somewhere`" doesn't exist.  Each time I created or modified a .organ file I needed to hand edit the .organ file to fix the `<path>`.   I'll see if I can put together a **reproducible** test

( In my environment, ${HOME}/GrandOrgue is a symbolic link to /tank/GOfiles/GrandOrgue (tank/GOfiles being a ZFS filesystem) this might be causing confusion with WX. )

larspalo replied:
Strange. I've used plenty of ODFs coming from OdfEdit that contain ../some/path and they all work fine in GoOdf when modified or just loaded.

kwhite-uottawa replied:
The good news is that, despite my best efforts, I've been **unable** to duplicate the error I was seeing in saving files!  It's probably safe to revert that change.  I'll revisit if I'm ever able to provide a simple test confirming the issue.

One thing I notice, though,  is that the path separator is different in the Attack and Release sections on loaded .organ files: Attack(s) have `/`, Release(s) `\`   (Related to the issue I was having?  Perhaps.)
![Slash](https://github.com/user-attachments/assets/1686d1c1-8115-44d1-a27a-d1d42b4693bf)

When you initially load the samples into the organ, both Attack(s) and Release(s) have `/` as path separator.
![Load](https://github.com/user-attachments/assets/5dd1586e-739a-4d0d-b8ce-813d9af2077a)

larspalo reviewed (changes requested):
I think that this PR should be somewhat broken down to individual PRs (move the whole testing and building for FreeBSD to another). There are some good things in this PR so I sure hope you don't feel disheartened by my comments!

kwhite-uottawa commented on .github/workflows/release.yml:
```diff
@@ -1,6 +1,6 @@
 name: Release
 
-on: [push]
+on: [workflow_dispatch]
```
This was definitely a mistake on my part.  This was not supposed to part of the PR.

kwhite-uottawa commented on src/GOODFFunctions.h:
```diff
@@ -74,14 +80,16 @@ namespace GOODF_functions {
 			if (theFile.FileExists()) {
 				return theFile.GetFullPath();
 			}
+			wxLogWarning("%s does not exist.  Removed from .organ file", relativePath);
```
I'm showing my age here. In pre-computer days, in English typewriter style, you ended a sentence with two spaces.  My fingers still follow this automatically!

kwhite-uottawa commented on src/GOODFFrame.cpp:
```diff
@@ -811,7 +811,7 @@ void GOODFFrame::DoOpenOrgan(wxString filePath) {
 	m_organTreeCtrl->SelectItem(tree_organ);
 	SetImportXfadeMenuItemState();
-	this->Raise();
+	// this->Raise();  should not be necessary
```
The reason was that if the LogWindow was being populated (and raised) then the Main window would in turn bury the LogWindow.  Adding a condition to Raise the LogWindow if it's been populated might be a better way to handle this.  I could try that.

larspalo replied:
> Adding a condition to Raise the LogWindow if it's been populated might be a better way to handle this. I could try that.

Yes, indeed. Adding something like

```
if (m_logWindow->GetFrame()->IsShown())
	m_logWindow->GetFrame()->Raise();
```

after line 814 could perhaps work better on all platforms.

kwhite-uottawa replied:
Perfect!  It shall be done.

kwhite-uottawa commented on src/Rank.cpp:
```diff
@@ -1100,6 +1100,29 @@ void Rank::addTremulantToPipes(
 
 		// if there are any matching attacks we add them
 		if (!pipeAttacksToAdd.IsEmpty()) {
+			// warn if all previous attacks ignore wave tremulant
```
If multiple attacks exist, then only a warning is shown.  If the bool `setSingle` is `false`, then no attack will be modified.  No method is provided to toggle `setSingle`, a user accessible check box will need to be added.

Another issue I was trying to solve was when there are less tremulant pipes than non-tremulant pipes.  This is the case in the Barton SoloStrings and SoloStringsTrem ranks I had.  I was hoping to avoid error prone hand editing to ignores-tremulant all the non-tremulant pipes without matching tremulant pipes.

I think that having at least a warning would be a good idea:  "Have you forgotten to select Load pipes to play when wave tremulant is off"
1114 becomes` if (!hasNonWave) {`
remove lines 1106, 1117-1125

larspalo replied:
> If multiple attacks exist, then only a warning is shown.

But how to know if it's intentional or not.

> I think that having at least a warning would be a good idea

Yes, that was what I suggested in my initial comment. One should parse **all** the attacks, and if not at least one of them has "play when wave tremulant is off" - then issue a warning (when loading the wave tremulant samples). Still this might be perfectly valid since the user can load the non-tremmed versions later. I outlined the recommended procedure above.

kwhite-uottawa replied:
Pushed the wrong button and lost my much more detailed response!  Oh well.  Older, wiser (jury's out on that!).

Fortunately GoOdf does not need to actually read the samples.  I have a script that generates a test sample-set tree (empty *.wav).  I'll look at coding a warning popup.

larspalo replied:
As I wrote in the README.md file:

" Always work on copies or on files you don't mind having destroyed!"

Anyway, thanks for trying to improve the program!

kwhite-uottawa replied:
I have a mental block.  Do you have an example of a real-world ODF that uses all three states of `isTremulant` (NULL, true, false) on a single pipe in a meaningful way?  A rank with a mix of invdividual pipes with `isTremulant` NULL, **or** `isTremulant` defined makes perfect sense.  All three states in a single pipe, not so much.

Here's a contrived single pipe with four attacks. Attacks 0 and 1 will always be selected,  with attack 2 when Tremulant is on, and attack 3 when Tremulant is off.

```
Pipe001=hwfake8\036.wav
Pipe001AttackCount=3
Pipe001Attack001=hwfake8\036_bis.wav
Pipe001Attack002=hwfake8_tr\036.wav
Pipe001Attack002IsTremulant=1
Pipe001Attack003=hwfake8_rel\036.wav
Pipe001Attack003IsTremulant=0
```

| Attack | Switch | IsTremulant | Played |
|--------|--------|--------|--------|
| 0 | ON | NULL | Yes |
| 1 | ON | NULL | Yes |
| 2 | ON | 1 | Yes |
| 3 | ON | 0 | No |
| 0 | OFF | NULL | Yes |
| 1 | OFF | NULL | Yes |
| 2 | OFF | 1 | No |
| 3 | OFF | 0 | Yes |

kwhite-uottawa replied:
I've pushed 382de96.  No attacks/releases are modified, and only one Dialog is displayed per rank update if there are any detected  isTremulant mixes.  More places where this could happen were uncovered in testing, so there are more lines changed than expected!

kwhite-uottawa replied:
@larspalo would you be able to run https://gist.github.com/kwhite-uottawa/4c1486e06a873c23d250d28228b0cf52#file-checktrem-sh on your .organ files?  It should be silent unless there are pipes with three tremulant states.  It's a script that runs awk to convert Pipe* lines to SQL which is then to be piped to sqlite3.

Output over the .organ files I have collected `find ../../Organs/ -name '*.organ' | sed -e 's+.*+./checktrem.sh "&" | sqlite3+' | sh -v 2>&1 | tee checktrem.log` is in https://gist.github.com/kwhite-uottawa/4c1486e06a873c23d250d28228b0cf52#file-checktrem-log

Output for suspect .organ files looks like this:
```
./checktrem.sh "../../Organs/Buckerburg/Audio/BuckerburgODF-trem.organ" | sqlite3
rank     name               pipeno  what     nsamples  nnulls  ndefs  toterrs
-------  -----------------  ------  -------  --------  ------  -----  -------
Stop016  37 Quintadena 16'  001     attack   2         1       1      300    
Stop016  37 Quintadena 16'  001     release  5         3       2      300    
Stop016  37 Quintadena 16'  002     attack   2         1       1      300    
Stop016  37 Quintadena 16'  002     release  5         3       2      300    
Stop016  37 Quintadena 16'  003     attack   2         1       1      300     
```

larspalo replied:
I've got no problem reported for most sample sets, nor did I expect it. Barton is a special case of course. Personally, I don't see much point in fussing about the tremulant states either. A simple warning, well phrased, if the case is detected is absolutely good enough for me.

kwhite-uottawa replied:
> I've got no problem reported for most sample sets, nor did I expect it. Barton is a special case of course. Personally, I don't see much point in fussing about the tremulant states either. A simple warning, well phrased, if the case is detected is absolutely good enough for me.

Barton is indeed a special case!  Doing the right thing for a beginning GoOdf user with that sample set can be a challenge.

I had removed updating the tremulant and only display a warning message in 382de9684d6ee6d44132c0153d8a34d84146e282 Better phrasing would be helpful.  I just don't have any ideas.
```
wxMessageDialog msg(::wxGetApp().m_frame, wxT("Warning: A mix of tremulant and non-tremulant pipes in a rank."), wxT("Pipe Tremulants Warning"), wxOK|wxCENTRE|wxICON_EXCLAMATION);
```

kwhite-uottawa commented on src/GOODFFunctions.h:
```diff
@@ -26,6 +26,12 @@
+#define SHOWLOGWINDOW do { \
+	wxLogWindow* w = ::wxGetApp().m_frame->GetLogWindow(); \
+	w->Show(true); \
+	w->GetFrame()->Raise(); \
```
In remote testing of a .organ file with all pipes missing, `w->GetFrame()->Raise();` is quite expensive and GoOdf appears hung if there are a lot of SHOW_LOG_WINDOW calls.  Since the LogWindow is now raised in 1f1178dd639f28a31996fc781bfd97f3cc3437da

larspalo commented on src/Rank.cpp:
```diff
@@ -797,6 +804,17 @@ void Rank::addToPipes(
 
 		// if there are any matching attacks we add them
 		if (!pipeAttacksToAdd.IsEmpty()) {
+
```
I think that all this (in more or less every place it's done) and the relevant macros perhaps would be better done as a public function of a Pipe. Then one could just ask the Pipe if it has conflicting isTremulant values in the attacks and save a few repeated lines. The function would just return a bool value which in turn would decide whether to trigger the warning message or not.

larspalo replied:
And possibly fire that message dialog itself as well...

kwhite-uottawa replied:
> And possibly fire that message dialog itself as well...

That had occurred to me too.  I'm not a C++ programmer.  Even though I certainly understand and agree with what you're saying (public function of Pipe, etc.) implementation might take me some figuring out.

larspalo replied:
Actually, as a message dialog would block the execution, it could be better for the flow to just log the warning and proceed.

larspalo commented on src/GOODFFunctions.h:
```diff
@@ -26,6 +26,11 @@
 #include <wx/filename.h>
 #include <vector>
 #include "GOODF.h"
+// macro to both show and raise the LogWindow
```
Revert this macro and all places where it's called as it currently doesn't do anything better than the existing triggering line.
